### PR TITLE
Implemente signup service object

### PR DESCRIPTION
**Pull request by eduardopoleo**

### CODE REVIEW

File: app/services/user_signup.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
+class UserSignup 
+  attr_accessor :user, :status, :error_message
+
+  def initialize(user)
+    @user = user
+  end 
+
+  def sign_up(stripe_token, invitation_token)
+    if user.valid?
+      handle_invitation(invitation_token)
```
Comment (brandoncc) [inline app/services/user_signup.rb]: In real life you might want to handle the invitation after the charge was successful.

File: spec/controllers/users_controller_spec.rb [hunk 2]
```diff
@@ -15,129 +15,36 @@
...
-    end
-
-    context 'with invalid user information input' do    
-      before {post :create, user: {password: 'password', full_name: 'eduardo'}}
-      it 'does not save the record' do
-        User.count.should == 0
+    context 'successful user signup' do
+      before do
+        response = double(:signup_result, status: :success)
+        UserSignup.any_instance.should_receive(:sign_up).and_return(response)
```
Comment (brandoncc) [inline spec/controllers/users_controller_spec.rb:21]: You should stick with the newer `expect` syntax.

``` ruby
expect_any_instance_of(UserSignup).to receive(:sign_up).and_return(response)
```

File: spec/services/user_signup_spec.rb [hunk 3]
```diff
@@ -0,0 +1,86 @@
+require 'spec_helper'
+
+describe UserSignup do
+  describe '#signup' do
+    context 'valid personal info and credit card' do
+      let(:charge) {double(:charge, successful?: true)}
+      before do
+        StripeWrapper::Charge.should_receive(:create).and_return(charge)
```
Comment (brandoncc) [inline spec/services/user_signup_spec.rb:8]: Same here, you should use the `expect` syntax.

``` ruby
expect(StripeWrapper::Charge).to receive(:create).and_return(charge)
```

File: spec/services/user_signup_spec.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      end
+        it 'creates a user ' do
+          UserSignup.new(Fabricate.build(:user)).sign_up("some_stripe_token", nil)
+          expect(User.count).to eq(1)
+        end
+
+        it 'makes the user follow the guest' do
+          alice = Fabricate(:user)
+          invitation = Fabricate(:invitation, user: alice, guest_email: 'joe@example.com')
+          UserSignup.new(Fabricate.build(:user, email: 'joe@example.com', password: 'password', full_name: "John Doe")).sign_up("some_stripe_token", invitation.token)
```
Comment (brandoncc) [inline spec/services/user_signup_spec.rb]: You should split long lines at roughly 80 characters.

File: spec/services/user_signup_spec.rb [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+        it 'sends out email with valid input' do
+          UserSignup.new(Fabricate.build(:user)).sign_up("some_stripe_token", nil)
+          ActionMailer::Base.deliveries.should_not be_empty
+        end
+
+        it 'sends out to the right recipient' do
+          UserSignup.new(Fabricate.build(:user)).sign_up("some_stripe_token", nil)
+          message =  ActionMailer::Base.deliveries.last
+          message.to.should == [User.first.email]
```
Comment (brandoncc) [inline spec/services/user_signup_spec.rb]: You used `should` again on most of the specs from here down.

File: spec/services/user_signup_spec.rb [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    context 'with invalid personal info' do
+      it 'does not save the record' do
+        UserSignup.new(Fabricate.build(:user, email: '')).sign_up("3424", nil)
+        User.count.should == 0
+      end
+
+      it 'does not charge the card' do
+        UserSignup.new(Fabricate.build(:user, email: '')).sign_up("3424", nil)
+        StripeWrapper::Charge.should_not_receive(:create)
```
Comment (brandoncc) [inline spec/services/user_signup_spec.rb:88]: I believe these lines should switch order. You don't eavesdrop on the method until you execute this line.

Comment (eduardopoleo) [inline spec/services/user_signup_spec.rb:88]: hey Brandon I'm not entirely sure what lines are you referring in here? The does not charge card spec?

**Pull request merged**